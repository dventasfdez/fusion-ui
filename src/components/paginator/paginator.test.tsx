import React from 'react';
import Paginator, {IPaginator} from './paginator';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import {fireEvent, render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const paginatorTest = (args: IPaginator) => {
  return <Paginator {...args} data-testid="paginator" />;
};

const paginatorTestWithoutTestId = (args: IPaginator) => {
  return <Paginator {...args} />;
};

test('renders and match snapshot', () => {
  const component = renderer.create(paginatorTest({pages: 20}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders and match snapshot without test id', () => {
  const component = renderer.create(paginatorTestWithoutTestId({pages: 20}));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders default paginator', () => {
  const {getByTestId} = render(paginatorTest({pages: 20}));
  const paginator = getByTestId('paginator');
  expect(paginator.querySelector('paginator')).toBeDefined();
});

test('renders collapsed paginator', () => {
  const {getByTestId} = render(paginatorTest({pages: 20, collapsed: true}));
  const paginator = getByTestId('paginator');
  expect(paginator.querySelector('paginator-collapsed')).toBeDefined();
});

test('incerment page default', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20}));
  const nextButton = getByTestId('paginator-button-next');
  expect(getByText(1)).toBeDefined();
  userEvent.click(nextButton);
  expect(getByText(6)).toBeDefined();
});

test('incerment page truncate', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20, truncate: 'right'}));
  const nextButton = getByTestId('paginator-button-next');
  expect(getByText(1)).toBeDefined();
  userEvent.click(nextButton);
  expect(getByText(6)).toBeDefined();
});

test('decrement page default', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20, defaultPage: 6}));
  const prevButton = getByTestId('paginator-button-previous');
  expect(getByText(7)).toBeDefined();
  userEvent.click(prevButton);
  expect(getByText(1)).toBeDefined();
});

test('decrement page truncated', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20, truncate: 'right'}));
  const nextButton = getByTestId('paginator-button-next');
  const prevButton = getByTestId('paginator-button-previous');
  userEvent.click(nextButton);
  expect(getByText(7)).toBeDefined();
  userEvent.click(prevButton);
  expect(getByText(1)).toBeDefined();
});

test('truncate to start (left)', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20, truncate: 'left'}));
  const nextChunckBtn = getByTestId('paginator-button-next');
  if (nextChunckBtn) fireEvent.click(nextChunckBtn);
  const trunc = getByTestId('paginator-truncate-left');
  userEvent.click(trunc);
  expect(getByText(1)).toHaveClass('paginator-page_active');
});

test('truncate to finish (right)', () => {
  const {getByText, getByTestId} = render(paginatorTest({pages: 20, truncate: 'right'}));
  const trunc = getByTestId('paginator-truncate-right');
  userEvent.click(trunc);
  expect(getByText(20)).toHaveClass('paginator-page_active');
});

test('number click', () => {
  const onChange = jest.fn();
  const {getByTestId} = render(paginatorTest({pages: 20, defaultPage: 3, onChangePage: onChange}));
  const buttonNum = getByTestId('paginator-button-num-1');
  userEvent.click(buttonNum);
  expect(onChange).toBeCalledWith(1);
});

test('jump option to start', () => {
  const {getByTestId, container} = render(
    paginatorTest({pages: 20, collapsed: true, defaultPage: 5, collapsedOptions: {jump: 'all'}})
  );
  const jump = getByTestId('paginator-button-jump-previous');

  userEvent.click(jump);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('1');
});

test('jump option to finish', () => {
  const {getByTestId, container} = render(
    paginatorTest({pages: 20, collapsed: true, defaultPage: 5, collapsedOptions: {jump: 'all'}})
  );
  const jump = getByTestId('paginator-button-jump-next');

  userEvent.click(jump);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('20');
});

test('collapsed arrow navigation left', () => {
  const {getByTestId, container} = render(
    paginatorTest({pages: 20, collapsed: true, defaultPage: 5, collapsedOptions: {jump: 'all'}})
  );
  const previous = getByTestId('paginator-button-previous');
  userEvent.click(previous);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('4');
});

test('collapsed arrow navigation right', () => {
  const {getByTestId, container} = render(
    paginatorTest({pages: 20, collapsed: true, defaultPage: 5, collapsedOptions: {jump: 'all'}})
  );
  const next = getByTestId('paginator-button-next');
  userEvent.click(next);
  expect(container.getElementsByClassName('input')[0]?.innerHTML).toBe('6');
});

test('select page from dropdown collapsed', () => {
  const {getByTestId} = render(paginatorTest({pages: 20, collapsed: true, defaultPage: 5, collapsedOptions: {jump: 'all'}}));
  const dropdown = getByTestId('paginator-dropdown');
  if (dropdown) userEvent.click(dropdown);
  const DItem = screen.getByTestId('paginator-dropdown-item-4');
  userEvent.click(DItem);
  expect(dropdown.getElementsByClassName('input')[0]?.innerHTML).toBe('4');
});

test('minimum pages render', () => {
  const {queryByText} = render(paginatorTest({pages: 5}));
  expect(queryByText(/navigate_next/i)).toBeNull();
});

test('disabled paginator', () => {
  const {getByTestId} = render(paginatorTest({pages: 20, disabled: true}));
  const disabledNumber = getByTestId('paginator-button-next');
  expect(disabledNumber).toHaveAttribute('disabled');
});

test('Render paginator without data-testid', () => {
  const {container} = render(<Paginator pages={20} />);
  expect(container).toMatchSnapshot();
});

test('Render truncate right paginator active on last element', () => {
  const {container} = render(<Paginator pages={20} truncate="right" active={20} />);
  expect(container).toMatchSnapshot();
});

test('Render truncate right paginator active on last element with data-testid', () => {
  const {container} = render(<Paginator pages={20} truncate="right" active={20} data-testid="test" />);
  expect(container).toMatchSnapshot();
});

test('Render truncate right paginator active on last element', () => {
  const {getByTestId} = render(<Paginator pages={20} truncate="right" defaultPage={20} data-testid="paginator" />);
  const previousChunckBtn = getByTestId('paginator-button-previous');
  if (previousChunckBtn) fireEvent.click(previousChunckBtn);
  const firstElement = getByTestId('paginator-truncate-right');
  expect(firstElement).toHaveClass('paginator-page_active');
});

test('Render truncate left paginator active on last element', () => {
  const {getByTestId, rerender} = render(<Paginator pages={20} truncate="left" defaultPage={0} data-testid="paginator" />);
  const nextChunckBtn = getByTestId('paginator-button-next');
  if (nextChunckBtn) fireEvent.click(nextChunckBtn);
  const firstElement = getByTestId('paginator-truncate-left');
  expect(firstElement).toHaveClass('paginator-page_active');
  const element = getByTestId('paginator-button-num-6');
  if (element) fireEvent.click(element);
  expect(element).toHaveClass('paginator-page_active');
  rerender(<Paginator pages={20} truncate="left" defaultPage={6} />);
  expect(() => getByTestId('paginator-truncate-left')).toThrow();
});

test('Render truncate left paginator active on first element', () => {
  const {container} = render(<Paginator pages={20} truncate="left" defaultPage={5} />);
  expect(container).toMatchSnapshot();
});

test('Render truncate left paginator active on first element with data-testid', () => {
  const {container} = render(<Paginator pages={20} truncate="left" defaultPage={5} data-testid="test" />);
  expect(container).toMatchSnapshot();
});

test('Render collapsed paginator with active !== 0', () => {
  const {container} = render(<Paginator pages={20} collapsed defaultPage={5} />);
  expect(container).toMatchSnapshot();
});

test('Render collapsed paginator with active !== 0 with collapsedOptions', () => {
  const {container} = render(<Paginator pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'left'}} />);
  expect(container).toMatchSnapshot();
});

test('Render collapsed paginator with active !== 0 with collapsedOptions', () => {
  const {container} = render(<Paginator pages={20} collapsed defaultPage={5} collapsedOptions={{jump: 'right'}} />);
  expect(container).toMatchSnapshot();
});
