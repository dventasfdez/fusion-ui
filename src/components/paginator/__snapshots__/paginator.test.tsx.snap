// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render collapsed paginator with active !== 0 1`] = `
<div>
  <div
    class="paginator-collapsed "
  >
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <div
      class="input-wrapper   paginator-collapsed-dropdown"
    >
      <div
        class="dropdown "
        data-show="false"
      >
        <div
          class="dropdown-button "
        >
          <div
            class="input-container"
          >
            <div
              class="input"
            >
              5
            </div>
            <span
              class="material-icons input-icon-box"
            >
              expand_more
            </span>
          </div>
        </div>
        <div
          class="dropdown-menu hidden "
          style="width: 0px;"
        >
          <ul>
            <li
              class="dropdown-item "
            >
              1
            </li>
            <li
              class="dropdown-item "
            >
              2
            </li>
            <li
              class="dropdown-item "
            >
              3
            </li>
            <li
              class="dropdown-item "
            >
              4
            </li>
            <li
              class="dropdown-item selected"
            >
              5
            </li>
            <li
              class="dropdown-item "
            >
              6
            </li>
            <li
              class="dropdown-item "
            >
              7
            </li>
            <li
              class="dropdown-item "
            >
              8
            </li>
            <li
              class="dropdown-item "
            >
              9
            </li>
            <li
              class="dropdown-item "
            >
              10
            </li>
            <li
              class="dropdown-item "
            >
              11
            </li>
            <li
              class="dropdown-item "
            >
              12
            </li>
            <li
              class="dropdown-item "
            >
              13
            </li>
            <li
              class="dropdown-item "
            >
              14
            </li>
            <li
              class="dropdown-item "
            >
              15
            </li>
            <li
              class="dropdown-item "
            >
              16
            </li>
            <li
              class="dropdown-item "
            >
              17
            </li>
            <li
              class="dropdown-item "
            >
              18
            </li>
            <li
              class="dropdown-item "
            >
              19
            </li>
            <li
              class="dropdown-item "
            >
              20
            </li>
          </ul>
        </div>
      </div>
      <input
        name="paginator-selector"
        readonly=""
        type="hidden"
        value="5"
      />
    </div>
    <span>
      of 
      20
       pages
    </span>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render collapsed paginator with active !== 0 with collapsedOptions 1`] = `
<div>
  <div
    class="paginator-collapsed "
  >
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        first_page
      </span>
    </button>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <div
      class="input-wrapper   paginator-collapsed-dropdown"
    >
      <div
        class="dropdown "
        data-show="false"
      >
        <div
          class="dropdown-button "
        >
          <div
            class="input-container"
          >
            <div
              class="input"
            >
              5
            </div>
            <span
              class="material-icons input-icon-box"
            >
              expand_more
            </span>
          </div>
        </div>
        <div
          class="dropdown-menu hidden "
          style="width: 0px;"
        >
          <ul>
            <li
              class="dropdown-item "
            >
              1
            </li>
            <li
              class="dropdown-item "
            >
              2
            </li>
            <li
              class="dropdown-item "
            >
              3
            </li>
            <li
              class="dropdown-item "
            >
              4
            </li>
            <li
              class="dropdown-item selected"
            >
              5
            </li>
            <li
              class="dropdown-item "
            >
              6
            </li>
            <li
              class="dropdown-item "
            >
              7
            </li>
            <li
              class="dropdown-item "
            >
              8
            </li>
            <li
              class="dropdown-item "
            >
              9
            </li>
            <li
              class="dropdown-item "
            >
              10
            </li>
            <li
              class="dropdown-item "
            >
              11
            </li>
            <li
              class="dropdown-item "
            >
              12
            </li>
            <li
              class="dropdown-item "
            >
              13
            </li>
            <li
              class="dropdown-item "
            >
              14
            </li>
            <li
              class="dropdown-item "
            >
              15
            </li>
            <li
              class="dropdown-item "
            >
              16
            </li>
            <li
              class="dropdown-item "
            >
              17
            </li>
            <li
              class="dropdown-item "
            >
              18
            </li>
            <li
              class="dropdown-item "
            >
              19
            </li>
            <li
              class="dropdown-item "
            >
              20
            </li>
          </ul>
        </div>
      </div>
      <input
        name="paginator-selector"
        readonly=""
        type="hidden"
        value="5"
      />
    </div>
    <span>
      of 
      20
       pages
    </span>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render collapsed paginator with active !== 0 with collapsedOptions 2`] = `
<div>
  <div
    class="paginator-collapsed "
  >
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <div
      class="input-wrapper   paginator-collapsed-dropdown"
    >
      <div
        class="dropdown "
        data-show="false"
      >
        <div
          class="dropdown-button "
        >
          <div
            class="input-container"
          >
            <div
              class="input"
            >
              5
            </div>
            <span
              class="material-icons input-icon-box"
            >
              expand_more
            </span>
          </div>
        </div>
        <div
          class="dropdown-menu hidden "
          style="width: 0px;"
        >
          <ul>
            <li
              class="dropdown-item "
            >
              1
            </li>
            <li
              class="dropdown-item "
            >
              2
            </li>
            <li
              class="dropdown-item "
            >
              3
            </li>
            <li
              class="dropdown-item "
            >
              4
            </li>
            <li
              class="dropdown-item selected"
            >
              5
            </li>
            <li
              class="dropdown-item "
            >
              6
            </li>
            <li
              class="dropdown-item "
            >
              7
            </li>
            <li
              class="dropdown-item "
            >
              8
            </li>
            <li
              class="dropdown-item "
            >
              9
            </li>
            <li
              class="dropdown-item "
            >
              10
            </li>
            <li
              class="dropdown-item "
            >
              11
            </li>
            <li
              class="dropdown-item "
            >
              12
            </li>
            <li
              class="dropdown-item "
            >
              13
            </li>
            <li
              class="dropdown-item "
            >
              14
            </li>
            <li
              class="dropdown-item "
            >
              15
            </li>
            <li
              class="dropdown-item "
            >
              16
            </li>
            <li
              class="dropdown-item "
            >
              17
            </li>
            <li
              class="dropdown-item "
            >
              18
            </li>
            <li
              class="dropdown-item "
            >
              19
            </li>
            <li
              class="dropdown-item "
            >
              20
            </li>
          </ul>
        </div>
      </div>
      <input
        name="paginator-selector"
        readonly=""
        type="hidden"
        value="5"
      />
    </div>
    <span>
      of 
      20
       pages
    </span>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        last_page
      </span>
    </button>
  </div>
</div>
`;

exports[`Render paginator without data-testid 1`] = `
<div>
  <div
    class="paginator "
  >
    <button
      class="paginator-icon disabled"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <button
      class="paginator-page_active"
      type="button"
    >
      1
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      2
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      3
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      4
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      5
    </button>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render truncate left paginator active on first element 1`] = `
<div>
  <div
    class="paginator "
  >
    <button
      class="paginator-icon disabled"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      1
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      2
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      3
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      4
    </button>
    <button
      class="paginator-page_active"
      type="button"
    >
      5
    </button>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render truncate left paginator active on first element with data-testid 1`] = `
<div>
  <div
    class="paginator "
    data-testid="test"
  >
    <button
      class="paginator-icon disabled"
      data-testid="test-button-previous"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-1"
      type="button"
    >
      1
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-2"
      type="button"
    >
      2
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-3"
      type="button"
    >
      3
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-4"
      type="button"
    >
      4
    </button>
    <button
      class="paginator-page_active"
      data-testid="test-button-num-5"
      type="button"
    >
      5
    </button>
    <button
      class="paginator-icon "
      data-testid="test-button-next"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render truncate right paginator active on last element 1`] = `
<div>
  <div
    class="paginator "
  >
    <button
      class="paginator-icon disabled"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <button
      class="paginator-page_active"
      type="button"
    >
      1
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      2
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      3
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      4
    </button>
    <button
      class="paginator-page"
      type="button"
    >
      5
    </button>
    <span
      class="paginator-truncate"
    >
      ...
    </span>
    <button
      class="paginator-page"
      type="button"
    >
      20
    </button>
    <button
      class="paginator-icon "
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`Render truncate right paginator active on last element with data-testid 1`] = `
<div>
  <div
    class="paginator "
    data-testid="test"
  >
    <button
      class="paginator-icon disabled"
      data-testid="test-button-previous"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_before
      </span>
    </button>
    <button
      class="paginator-page_active"
      data-testid="test-button-num-1"
      type="button"
    >
      1
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-2"
      type="button"
    >
      2
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-3"
      type="button"
    >
      3
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-4"
      type="button"
    >
      4
    </button>
    <button
      class="paginator-page"
      data-testid="test-button-num-5"
      type="button"
    >
      5
    </button>
    <span
      class="paginator-truncate"
    >
      ...
    </span>
    <button
      class="paginator-page"
      data-testid="test-truncate-right"
      type="button"
    >
      20
    </button>
    <button
      class="paginator-icon "
      data-testid="test-button-next"
      type="button"
    >
      <span
        class="material-icons"
      >
        navigate_next
      </span>
    </button>
  </div>
</div>
`;

exports[`renders and match snapshot 1`] = `
<div
  className="paginator "
  data-testid="paginator"
>
  <button
    className="paginator-icon disabled"
    data-testid="paginator-button-previous"
    onClick={[Function]}
    type="button"
  >
    <span
      className="material-icons"
    >
      navigate_before
    </span>
  </button>
  <button
    className="paginator-page_active"
    data-testid="paginator-button-num-1"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="paginator-page"
    data-testid="paginator-button-num-2"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="paginator-page"
    data-testid="paginator-button-num-3"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="paginator-page"
    data-testid="paginator-button-num-4"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="paginator-page"
    data-testid="paginator-button-num-5"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="paginator-icon "
    data-testid="paginator-button-next"
    onClick={[Function]}
    type="button"
  >
    <span
      className="material-icons"
    >
      navigate_next
    </span>
  </button>
</div>
`;

exports[`renders and match snapshot without test id 1`] = `
<div
  className="paginator "
>
  <button
    className="paginator-icon disabled"
    onClick={[Function]}
    type="button"
  >
    <span
      className="material-icons"
    >
      navigate_before
    </span>
  </button>
  <button
    className="paginator-page_active"
    onClick={[Function]}
    type="button"
  >
    1
  </button>
  <button
    className="paginator-page"
    onClick={[Function]}
    type="button"
  >
    2
  </button>
  <button
    className="paginator-page"
    onClick={[Function]}
    type="button"
  >
    3
  </button>
  <button
    className="paginator-page"
    onClick={[Function]}
    type="button"
  >
    4
  </button>
  <button
    className="paginator-page"
    onClick={[Function]}
    type="button"
  >
    5
  </button>
  <button
    className="paginator-icon "
    onClick={[Function]}
    type="button"
  >
    <span
      className="material-icons"
    >
      navigate_next
    </span>
  </button>
</div>
`;
